### PR TITLE
Adding support for convenience type overloads

### PR DESCRIPTION
**Pull request by spfink**
## Description

Adds support for convenience type overloads as we had in v1. This allows us to change API models to add custom types for any shape member. In this case, we are adding the Tagging object to S3 instead of requiring encoded strings. Also brings back support for the StateMachine type in StepFunctions.

## Testing
Updated existing tests to cover new changes.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by spfink** [comment]
Forgot about the new changelog stuff. Will add entry before pushing.

**Comment by dagnir** [comment]
When I did the initial move from FTL to poet for models, we removed this feature and said we'd revisit in the future. Do we want to have a more broad discussion around this?

I think one issue is that the internal representation and getter are still `String` right? Seems like it wouldn't be too user friendly if the customer wanted to modify a PutObjectRequest and modify the tag set after the fact.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AccessorsFactory.java [hunk 1]
```diff
@@ -56,6 +58,14 @@ public MethodSpec beanStyleGetter(MemberModel memberModel) {
         return new NonCollectionSetters(intermediateModel, shapeModel, memberModel, typeProvider).fluentDeclarations(returnType);
     }
 
+    public List<MethodSpec> convenienceSetterDeclarations(MemberModel memberModel, TypeName returnType) {
+        return intermediateModel.getCustomizationConfig().getConvenienceTypeOverloads().stream()
+                                .filter(c -> c.accepts(shapeModel, memberModel))
+                                .map(s -> new NonCollectionSetters(intermediateModel, shapeModel, memberModel, typeProvider)
+                                     .convenienceDeclaration(returnType, s))
```
Comment (kiiadi) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/AccessorsFactory.java]: Something funny going on with the indenting here.

**Comment by spfink** [comment]
@dagnir sure we can.

We can also easily change to have the adapters go both way so we can have getters for both.

**Comment by shorea** [comment]
Getters are somewhat problematic when it comes to naming. May make sense to make this a property of the customization on whether it generates a getter and what it's named.

**Pull request merged**